src/type_inf.jl

```diff
@@ -32,42 +32,51 @@ end
 
 function infer_type_assignment_rhs(binding, state, scope)
     rhs = binding.val.args[2]
-    if CSTParser.is_func_call(rhs)
-        callname = CSTParser.get_name(rhs)
-        if isidentifier(callname)
-            resolve_ref(callname, scope, state)
-            if hasref(callname)
-                rb = get_root_method(refof(callname), state.server)
-                if (rb isa Binding && (rb.type == CoreTypes.DataType || rb.val isa SymbolServer.DataTypeStore)) || rb isa SymbolServer.DataTypeStore
-                    settype!(binding, rb)
+    if is_loop_iter_assignment(binding.val)
+        settype!(binding, infer_eltype(rhs))
+    elseif headof(rhs) === :ref && length(rhs.args) > 1
+        ref = refof_maybe_getfield(rhs.args[1])
+        if ref isa Binding && ref.val isa EXPR
+            settype!(binding, infer_eltype(ref.val))
+        end
+    else
+        if CSTParser.is_func_call(rhs)
+            callname = CSTParser.get_name(rhs)
+            if isidentifier(callname)
+                resolve_ref(callname, scope, state)
+                if hasref(callname)
+                    rb = get_root_method(refof(callname), state.server)
+                    if (rb isa Binding && (CoreTypes.isdatatype(rb.type) || rb.val isa SymbolServer.DataTypeStore)) || rb isa SymbolServer.DataTypeStore
+                        settype!(binding, rb)
```

Comment:
Can we end up with abstract types here? If so, is that a problem?

Reply:
This is all slightly dodgy - it assumes datatype methods _are_ constructors that return an instance. Clearly there's no requirement for this but would for the vast majority of cases be true

Reply:
Yeah, that seems like a good assumption to make.